virtual_rainforest/core/readers.py

```diff
@@ -0,0 +1,119 @@
+so that new functions that implement data loading for a given file format can be
+added to those supported by :meth:`~virtual_rainforest.core.data.Data.load_from_file`.
+
+New file format readers are made available using the
+:func:`~virtual_rainforest.core.data.register_file_format_loader` decorator, which needs
+to specify the file formats supported (as a tuple of file suffixes) and then decorate a
+function that returns a :class:`~xarray.DataArray` suitable for validation using
+:func:`~virtual_rainforest.core.axes.validate_dataarray`. For
+example:
+
+.. code-block:: python
+
+    @register_file_format_loader(('.tif', '.tiff'))
+    def new_function_to_load_tif_data(...):
+        # code to turn tif file into a data array
+"""  # noqa: D205
+
+from pathlib import Path
+from typing import Callable
+
+from xarray import DataArray, load_dataset
+
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+
+FILE_FORMAT_REGISTRY: dict[str, Callable] = {}
+"""A registry for different file format loaders
+
+This dictionary maps a tuple of file format suffixes onto a function that allows the
+data to be loaded. That loader function should coerce the data into an xarray DataArray.
+
+Users can register their own functions to load from a particular file format using the
+:func:`~virtual_rainforest.core.data.register_file_format_loader` decorator. The
+function itself should have the following signature:
+
+    func(file: Path, file_var: str) -> DataArray
+"""
+
+
+def register_file_format_loader(file_types: tuple[str]) -> Callable:
+    """Adds a data loader function to the data loader registry.
+
+    This decorator is used to register a function that loads data from a given file type
+    and coerces it to a DataArray.
+
+    TODO - How to ensure that grids are registered before this decorator is used.
+
+    Args:
+        file_types: A tuple of strings giving the file type that the function will map
+            onto the Grid. The strings should match expected file suffixes for the file
+            type.
+    """
+
+    def decorator_file_format_loader(func: Callable) -> Callable:
+
+        # Ensure file_type is an iterable
+        if isinstance(file_types, str):
+            _file_types = (file_types,)
+        else:
+            _file_types = file_types
+
+        # Register the mapper function for each combination of grid type and file type
+        for this_ft in _file_types:
+
+            if this_ft in FILE_FORMAT_REGISTRY:
+                LOGGER.debug(
+                    "Replacing existing data loader function for %s",
+                    this_ft,
+                )
+            else:
+                LOGGER.debug(
+                    "Adding data loader function for %s",
+                    this_ft,
+                )
+
+            FILE_FORMAT_REGISTRY[this_ft] = func
+
+        return func
+
+    return decorator_file_format_loader
+
+
+@register_file_format_loader(file_types=(".nc",))
```

Comment:
Is usage of `@register_file_format_loader` documented anywhere?

Reply:
It is documented in the `readers.py` docstring, but not outside of the API as it is a developer tool.

Reply:
Ahh yeah missed that sorry!